FEATURE: show status on chat mentions when user haven't used autocomplete

AndrewPrigorshnev opened the pull request:
We added user status to inline mentions in chat in d4a5b795:

<img width="335" alt="Screenshot 2023-06-09 at 00 20 45" src="https://github.com/discourse/discourse/assets/1274517/e73c768c-e264-4dc2-806c-8a070158401d">

That solution relies on the next behavior. When a user starts typing a mention, the client loads users from the server and shows them in the autocompletion popup:

<img width="410" alt="Screenshot 2023-06-09 at 00 14 07" src="https://github.com/discourse/discourse/assets/1274517/46f492e4-5cce-47a4-8d0c-1075aaeb5e2e">

When the user applies the autocomplete suggestion, we add the chosen user to the collection of mentioned users on the message model, which enables us to show user status on the mention of that user.

The problem is that sometimes users may ignore autocomplete suggestions and just keep typing, or they may paste a message with mentions to the composer from the clipboard. In such cases, we don't have user models at hand to take user statuses from them. This PR handles such cases by sending an additional request to the server.

That additional request:
- is only sent if users weren't loaded before via the autocompletion popup
- is sent _before_ sending the main request that posts the message, this way everything looks smoothly (If we sent it after or alternatively deliver users via message bus, when posting the message, it would first appear without status on mentions and then in a moment the status on mentions would appear, this doesn't look well)
- if it fails for some reason, we ignore the exception to not interrupt the more important request that posts the message. Mentions will be without status in this case (thought status will appear after reloading the page), but that's better than ruining posting the message.

jjaffeux commented on app/assets/javascripts/discourse/app/lib/collections.js:
```diff
@@ -0,0 +1,17 @@
+  For example:
+  except(["a", "b", "c", "d"], ["a", "b"]) returns ["c", "d"]
+ */
+export function except(left, right) {
+  const result = [];
+  const rightSet = new Set(right);
+  for (const item of left) {
+    if (!rightSet.has(item)) {
+      result.push(item);
+    }
+  }
+  return result;
+}
```
Any opinion on adding this @CvX ?

CvX replied:
I see it's used to get the difference between array of mentions in a single chat message and an array of loaded users. The first array has the max size of 10 and usually is empty. `a.filter(x => !b.includes(x))` is good enough for that

AndrewPrigorshnev replied:
Oh, I actually forgot to remove a restriction here, I wanted to remove a call to `slice`, so instead of 

```javascript
  get #notLoadedMentions() {
    const parsed = parseMentionedUsernames(this._cooked).slice(
      0,
      this.siteSettings.max_mentions_per_chat_message
    );
    const loaded = [...this.mentionedUsers.values()].map((u) => u.username);
    return except(parsed, loaded);
  }
```

we have just

```javascript
  get #notLoadedMentions() {
    const parsed = parseMentionedUsernames(this._cooked);
    const loaded = [...this.mentionedUsers.values()].map((u) => u.username);
    return except(parsed, loaded);
  }
```

This needs an explanation. The thing is that `max_mentions_per_chat_message` is mostly used for restricting amount of users to be notified and that enforced on the server. But if you mention more users than allowed, the mention in the chat message will still look like a mention, so we should show user status on it, otherwise it will be misleading. The default value for this setting is 5, and if I mention 6 users I will be shown a popup that users won't be notified:

<img width="500" alt="Screenshot 2023-06-21 at 17 23 57" src="https://github.com/discourse/discourse/assets/1274517/0fe18728-da7f-4a42-8e89-1905c797c91c">

that's great, but the message will look like this, the sixth user won't have status, while in fact he has one:

<img width="500" alt="Screenshot 2023-06-21 at 17 24 03" src="https://github.com/discourse/discourse/assets/1274517/a017ae58-7682-4a5f-ade6-91303314e766">

So we need to remove a restriction here and pass as many usernames as was mentioned to `except`. From this point of view, looks like it make sense to have this except function instead of using `a.filter(x => !b.includes(x))` that has N^2 complexity?

I've just pushed a commit that removed a restriction. @jjaffeux may you have a look too? Note that this shouldn't lead to problems on the server – it's only one simple query on the server side, and if a user tries to mention a lot of users they either exceed the maximum message size or Rails will decline the request because the query string is too long, otherwise the server should be able to serve the request in reasonable time.

AndrewPrigorshnev replied:
We have to make a tricky choice here – show status on all mentions, but in this case sometime we may need to load a lot of users with status, or forbid it, but then status on mentions may be misleading.

jjaffeux replied:
I think in this case we don't really care about the "slower" perf function given it's called a limited amount of times on a very limited amount of elements, or am I missing something?

AndrewPrigorshnev replied:
I've pushed a commit that remove the `except` and uses `a.filter(x => !b.includes(x))` instead.

jjaffeux commented on plugins/chat/assets/javascripts/discourse/components/chat-channel.js:
```diff
@@ -1079,4 +1088,13 @@ export default class ChatLivePane extends Component {
+
+  async #tryToEnsureMentionsLoaded(message) {
+    try {
+      await message.ensureMentionsLoaded();
```
How about we add a param `ensureMentionsLoaded`, something like `ignore_failure`, `false` by default, that way we don't have to repeat that code everywhere and we could also remove the comment.

```suggestion
      await message.ensureMentionsLoaded();
```

AndrewPrigorshnev replied:
A nice suggestion 👍, done.

jjaffeux commented on plugins/chat/assets/javascripts/discourse/components/chat-thread.js:
```diff
@@ -254,6 +254,8 @@ export default class ChatThreadPanel extends Component {
 
     this.chatThreadPane.sending = true;
+
+    await message.ensureMentionsLoaded({ ignoreFailure: true });
```
do we need to await or could we just throw it and hope for the best?

jjaffeux commented on app/controllers/users_controller.rb:
```diff
@@ -1900,6 +1900,11 @@ def user_menu_messages
            }
   end
 
+  def lookup
+    users = User.where(username: params[:usernames])
+    render json: serialize_found_users(users)
+  end
+
```
```
import userSearch from "discourse/lib/user-search";
```


Was not working out for you?

AndrewPrigorshnev replied:
It wasn't, `userSearch` takes a search term (a part of a username) and returns found users. While in this PR I have a list of usernames, for example ["user1", "user2", "moderator", "admin"] and I need to look up and load those users.

jjaffeux commented on plugins/chat/assets/javascripts/discourse/models/chat-message.js:
```diff
@@ -239,6 +243,28 @@ export default class ChatMessage {
+        this.incrementVersion();
+      }
+    } catch (e) {
+      if (ignoreFailure) {
+        // eslint-disable-next-line no-console
+        console.warn("Cannot load mentioned users", e);
+      } else {
+        throw e;
+      }
```
We always call it with `ignoreFailure: true` so I would make it the default.

ZogStriP replied:
I agree with that one, if `ignore failure` is the happy path, it should be the default

AndrewPrigorshnev replied:
This was added after this suggestion https://github.com/discourse/discourse/pull/21935#discussion_r1235445298. When addressing that, I didn't notice we're ending up with an argument we effectively don't use, my bad.

I fixed it in b31580ec08add346c4c5d12eb5301e744dd371c8, I switched it back to the previous version, but made it simpler to address @jjaffeux's concern in that old comment.

jjaffeux commented on spec/requests/users_controller_spec.rb:
```diff
@@ -5225,6 +5225,84 @@ def users_found
+      expect(users[0]["id"]).to be(user_1.id)
+    end
+
+    it "returns users with ids, names and avatar_templates" do
+      get "/u/lookup/users.json", params: { usernames: [user_1.username] }
+
+      expect(response.status).to eq(200)
+      users = response.parsed_body["users"]
+
+      expect(users.length).to be(1)
+      expect(users[0]["id"]).to be(user_1.id)
+      expect(users[0]["username"]).to eq(user_1.username)
+      expect(users[0]["name"]).to eq(user_1.name)
+      expect(users[0]["avatar_template"]).to eq(user_1.avatar_template)
+    end
```
I would also check we don't have OTHER attributes, like password for example, that would prevent any mistake in changing serializers.

jjaffeux commented on app/controllers/users_controller.rb:
```diff
@@ -1900,6 +1901,11 @@ def user_menu_messages
            }
   end
 
+  def lookup
+    users = User.where(username: params[:usernames])
+    render json: serialize_found_users(users)
+  end
```
We need a limit here

jjaffeux replied:
Did you think about this?

AndrewPrigorshnev replied:
I've limited it to 20 users max ans also tested it in 0233046df11ed7ff1be15098377338f9b8f678b1.

jjaffeux commented:
Sorry finally got to trying it, made few comments which should be easy to fix and then we can merge.

nbianca commented on plugins/chat/test/javascripts/acceptance/user-status-on-mentions-test.js:
```diff
@@ -229,7 +234,7 @@ acceptance("Chat | User status on mentions", function (needs) {
     await click(".chat-message-expand");
 
-    loggedInUser().appEvents.trigger("user-status:changed", {
+    loggedInUser().appEvents.triggedr("user-status:changed", {
```
This does not look right.

ZogStriP commented on app/assets/javascripts/discourse/app/lib/parse-mentions.js:
```diff
@@ -0,0 +1,9 @@
+export function parseMentionedUsernames(cooked) {
+  const html = new DOMParser().parseFromString(cooked, "text/html");
```
Is this every going to be used on the server-side? I think we can't do `DOMParser` there, right?

AndrewPrigorshnev replied:
This makes sense, and it's fixed now. We don't use `DOMParser` anymore (it was fixed in main in 73781c8a96e35062b2631263bd6afd37f7f7016b and then merged into this branch).

ZogStriP commented on app/controllers/users_controller.rb:
```diff
@@ -1912,6 +1913,11 @@ def user_menu_messages
   end
 
+  def lookup
+    users = User.where(username: params[:usernames])
```
Is there any way this could be abused to list users that someone else shouldn't be able to see?

Any rate limits?

AndrewPrigorshnev replied:
We go with default rate limits here, also this method is available only to authorized users, which is I believe what we do with this kind of methods by default. 

> No limits on the number of usernames passed by the user?

I've limited it to 20 users in 0233046df11ed7ff1be15098377338f9b8f678b1.

jjaffeux commented on app/assets/javascripts/discourse/tests/unit/lib/parse_mentions-test.js:
```diff
@@ -0,0 +1,23 @@
+
+    test("returns an empty array if there are no mentions", function (assert) {
+      const mentions = parseMentionedUsernames("<p>No mentions here</p>");
+      assert.equal(mentions.length, 0);
```
```suggestion
      assert.strictEqual(mentions.length, 0);
```

AndrewPrigorshnev replied:
This got outdated, I've deleted this test and the function it was testing.

jjaffeux commented on plugins/chat/assets/javascripts/discourse/models/chat-message.js:
```diff
@@ -94,6 +95,10 @@ export default class ChatMessage {
     if (args.thread) {
       this.thread = args.thread;
     }
+
+    this.site = getOwner(this).lookup("service:site");
+    this.siteSettings = getOwner(this).lookup("service:site-settings");
+    this.usersApi = getOwner(this).lookup("service:usersApi");
```
I wonder if we need to instantiate for each message if we are not going to use it each time

ZogStriP commented on plugins/chat/assets/javascripts/discourse/models/chat-message.js:
```diff
@@ -258,6 +262,28 @@ export default class ChatMessage {
+    } catch (e) {
+      if (ignoreFailure) {
+        // eslint-disable-next-line no-console
+        console.warn("Cannot load mentioned users", e);
```
Is there even a point in logging this in the console? 🤔

jjaffeux commented on spec/requests/users_controller_spec.rb:
```diff
@@ -5273,6 +5273,84 @@ def users_found
+    it "returns users without user status if status is disabled in settings" do
+      SiteSetting.enable_user_status = false
+      user_1.set_status!("off to dentist", "tooth")
+      user_1.reload
```
Is there any way we can avoid having to do this reload? This kind of knowledge is painful when we write specs.

AndrewPrigorshnev replied:
Oh, actually turned out we don't need to do it anymore, so I removed this `reload` call (see 7bee17f60dbfc26ee06f080d4f9f8bd2bfd76d67). We needed to do it in the past for some reason, but now it's fixed.

ZogStriP commented on spec/requests/users_controller_spec.rb:
```diff
@@ -5273,6 +5273,84 @@ def users_found
+      users = response.parsed_body["users"]
+      expect(users.length).to be(3)
+
+      ids = users.map { |user| user["id"] }
+      expect(ids).to include(user_1.id)
+      expect(ids).to include(user_2.id)
+      expect(ids).to include(user_3.id)
```
Maybe use `contrains_exactly` instead of matching ids one by one?

AndrewPrigorshnev replied:
Fixed in 145c787f96d748030ce0cba9e3301db1c9b91be7.

nbianca reviewed (commented):
Overall, I find it strange that we store details of mentioned users as a property of chat messages while they are shared between other parts of Discourse such as posts including other chat messages.

Even if it is hacky, I think I like the composer's approach which has a global store that contains the details of all seen users is slightly better. Another advantage of that is that it happens async... so the message would be rendered first and then if the mention is found and some details are updated the DOM element is updated.

CvX commented on app/assets/javascripts/discourse/app/services/users-api.js:
```diff
@@ -0,0 +1,13 @@
+import Service from "@ember/service";
+import { ajax } from "discourse/lib/ajax";
+import User from "discourse/models/user";
+
+export default class UsersApiService extends Service {
+  async lookupUsers(usernames) {
+    const response = await ajax(`/u/lookup/users.json`, {
+      type: "GET",
+      data: { usernames },
+    });
+    return response.users.map((u) => User.create(u));
+  }
```
Using a Service for fetching stuff definitely beats static methods on a class so that's a win 😃

jjaffeux replied:
Yes we decided to do it, mostly I wanted to discuss it with others, and not just push chat patterns everywhere

CvX commented on app/assets/javascripts/discourse/app/services/users-api.js:
```diff
@@ -0,0 +1,13 @@
+export default class UsersApiService extends Service {
+  async lookupUsers(usernames) {
+    const response = await ajax(`/u/lookup/users.json`, {
+      type: "GET",
```
```suggestion
```

GET requests are the default so no need to add it explicitly

CvX commented on app/assets/javascripts/discourse/app/services/users-api.js:
```diff
@@ -0,0 +1,13 @@
+      type: "GET",
+      data: { usernames },
+    });
+    return response.users.map((u) => User.create(u));
```
Using the static `create` method on a model is to be avoided. Use the store service instead

```js
@service store;
```

```suggestion
    return response.users.map((u) => this.store.createRecord("user", u));
```

(btw. note to self and modernize-js team/@davidtaylorhq - we should update the store so it can also accept a class instead of a string 😃)

ZogStriP commented:
@AndrewPrigorshnev what's the status on this?

AndrewPrigorshnev commented:
> @AndrewPrigorshnev what's the status on this?

At this moment I addressed all comments, I only need to answer on every comment with the things I explained when we were discussing this PR on the call (for further readers of this PR) and also fix some failing specs. I'm working on this now and will ping you a bit later.

AndrewPrigorshnev commented:
> Overall, I find it strange that we store details of mentioned users as a property of chat messages while they are shared between other parts of Discourse such as posts including other chat messages.
> 
> Even if it is hacky, I think I like the composer's approach which has a global store that contains the details of all seen users is slightly better.

We do like this in other places too, and in my opinion the main reason to do it like this is that it's the easiest way to make it simple and reliable. Designs that use shared state are generally very hard to get right. Using Ember Data or another similar thing would be a good solution here I think, this way we'll have the only solution for the whole app, and that'll be also a well-designed and well tested solution. While, writing a dedicated solution with shared state for different parts of the app either will be unreliable or will take too much time to implement it correctly.

> Another advantage of that is that it happens async... so the message would be rendered first and then if the mention is found and some details are updated the DOM element is updated.

Oh, we're actually trying to avoid such async rendering in chat because it doesn't look great. For example, see https://github.com/discourse/discourse/pull/23084.

AndrewPrigorshnev commented:
@ZogStriP @jjaffeux all the comments are addressed and answered now, so if there are no more objections, could you approve it?

jjaffeux commented on plugins/chat/assets/javascripts/discourse/models/chat-message.js:
```diff
@@ -96,6 +95,12 @@ export default class ChatMessage {
+    this.siteSettings = getOwnerWithFallback(this).lookup(
+      "service:site-settings"
+    );
+    this.usersApi = getOwnerWithFallback(this).lookup("service:usersApi");
```
```suggestion
    this.usersApi = getOwnerWithFallback(this).lookup("service:users-api");
```

jjaffeux commented on plugins/chat/test/javascripts/acceptance/user-status-on-mentions-test.js:
```diff
@@ -281,6 +286,61 @@ acceptance("Chat | User status on mentions", function (needs) {
+    );
+    await click(".chat-composer-button.-send");
+
+    loggedInUser().appEvents.trigger("user-status:changed", {
```
we should avoid to access a service through another object

jjaffeux commented on app/controllers/users_controller.rb:
```diff
@@ -2118,9 +2125,13 @@ def render_invite_error(message)
   def serialize_found_users(users)
-    each_serializer =
-      SiteSetting.enable_user_status? ? FoundUserWithStatusSerializer : FoundUserSerializer
+    {
+      users:
+        ActiveModel::ArraySerializer.new(users, each_serializer: found_user_serializer).as_json,
+    }
+  end
 
-    { users: ActiveModel::ArraySerializer.new(users, each_serializer: each_serializer).as_json }
+  def found_user_serializer
+    SiteSetting.enable_user_status? ? FoundUserWithStatusSerializer : FoundUserSerializer
   end
```
Im not sure "FoundUser" means a lot here? Can we return a not found user anyways?

---

Also Im not sure why we need to have the FoundUserWithStatusSerializer separate, and not just add the content in FoundUserSerializer as it's driven by a site setting, if we need finer control we can have it as an option of the serializer:

```ruby
ActiveModel::ArraySerializer.new(users, each_serializer: FoundUserSerializer, include_status: ...).as_json,
```

jjaffeux commented on plugins/chat/assets/javascripts/discourse/components/chat-channel.js:
```diff
@@ -550,6 +552,8 @@ export default class ChatChannel extends Component {
 
     resetIdle();
 
+    await message.tryToEnsureMentionsLoaded();
```
I don't think we need `tryTo` in the name, in essence any function is a try. What we are really doing is: ensuring mentions are loaded.
